packages/core/test/instruments.spec.ts

```diff
@@ -0,0 +1,95 @@
+import { envelop, Instruments, Plugin, useEngine } from '@envelop/core';
+
+describe('instruments', () => {
+  it('should instrument all graphql phases', async () => {
+    const result: string[] = [];
+    const make = (name: string): Plugin => ({
+      instruments: {
+        init: (_, w) => {
+          result.push(`pre-init-${name}`);
+          expect(w()).toBeUndefined();
+          result.push(`post-init-${name}`);
+        },
+        parse: (_, w) => {
+          result.push(`pre-parse-${name}`);
+          expect(w()).toBeUndefined();
+          result.push(`post-parse-${name}`);
+        },
+        validate: (_, w) => {
+          result.push(`pre-validate-${name}`);
+          expect(w()).toBeUndefined();
+          result.push(`post-validate-${name}`);
+        },
+        context: (_, w) => {
+          result.push(`pre-context-${name}`);
+          expect(w()).toBeUndefined();
+          result.push(`post-context-${name}`);
+        },
+        execute: async (_, w) => {
+          result.push(`pre-execute-${name}`);
+          expect(await w()).toBeUndefined();
+          result.push(`post-execute-${name}`);
+        },
+        subscribe: async (_, w) => {
+          result.push(`pre-subscribe-${name}`);
+          expect(await w()).toBeUndefined();
+          result.push(`post-subscribe-${name}`);
+        },
+      },
+    });
+
+    const getEnveloped = envelop({
+      plugins: [
+        useEngine({
+          execute: () => {
+            result.push('execute');
+            return new Promise(r => setTimeout(() => r('test'), 10));
+          },
+          subscribe: () => {
+            result.push('subscribe');
+            return new Promise(r => setTimeout(() => r('test'), 10));
+          },
+          parse: () => {
+            result.push('parse');
+            return { test: 'foo' };
+          },
+          validate: () => {
+            result.push('validate');
+            return 'test';
+          },
+        }),
+        make('1'),
+        make('2'),
+        make('3'),
```

Comment:
@enisdenjo Here the plugins are used as is, without manual composition.

You can see in the expect later that I'm checking that the order is correct and reflect the same order than the array.